### DISCUSSION TITLE
Automatische Phasenumschaltung konfigurieren

### DISCUSSION THREAD

**Discussion by protree** [goecharger/go-eCharger-API-v2#238] [General]
Hallo zusammen,

gibt es aktuell die Möglichkeit die Logik der automatischen Phasenumschaltung im Eco Modus zu beeinflussen, die über die Einstellungen in der Go-e App hinausgehen?

Ich würde gerne die Zeiten ändern, nach denen eine Phasenumschaltung stattfindet.

Spielen die API Schlüssel "psh", "mpwst", "mptwt" eine Rolle und werden diese in der aktuellen Firmware berücksichtigt? Wenn ja, was bewirken diese genau (ich werde leider aus der Beschreibung nicht ganz schlau)? Gibt es andere relevante Datenpunkte?

Kann man darüber hinaus die Zeit beeinflussen, nach der bei Unterschreitung der überschüssigen PV-Leistung die Ladung gestoppt wird?

**Comment by GuraxUbrax** [comment]
Ich versuche auch gerade, das einzurichten. 
Es gibt versteckte/deaktivierte Entitäten "Min. Phasenwechsel Zeit", "Min. Länge Ladepause", "Min. Ladezeit" und "Min. Ladeintervall". 
Mal sehen, wie sich das auswirkt, wenn man hier Werte setzt. In der App konnte ich diese Einstellungen jedenfalls nicht finden. 
Standardmäßig waren bei mir diese so eingestellt: Min. Phasenwechsel steht auf 10 Minuten, Min. Ladezeit auf 5. Die anderen auf 0.

**Reply by GuraxUbrax** [reply]
Ich rede von Home Assistant, anscheinend ist diese API übergreifend?!?

**Reply by GuraxUbrax** [reply]
OK, falsches Forum, ich rede tatsächlich von der HA-http-Integration der API (https://github.com/marq24/ha-goecharger-api2). 
Und diese liest die Standard-Werte wie oben genannt aus. 
Das Setzen eines Wertes scheint auch geklappt zu haben, nach Neustart ist der neue Wert noch vorhanden. 
Jetzt werde ich bei Gelegenheit mal testen, ob er auch wirklich diese minimalen Zeiten einhält (zumindest minimale Ladepause, minimale Ladezeit und minimale Zeit bis Phasenwechsel, den letzten Wert minimales Ladeintervall verstehe ich nicht).

**Reply by protree** [reply]
Ich denke, dass die HA Integration letztlich auch die API benutzt. Wahrscheinlich heißen die Werte nur anders. Hattest du Erfolg?

**Reply by GuraxUbrax** [reply]
Also, ich hatte es kurz mal probiert und es scheint auch zu funktionieren. Er hat z.B. auch eine längere Ladepause dann gemacht, als ich das eingestellt hatte. 
Aber dann dachte ich mir, was bringt mir z.B. eine eingestellte, minimale Ladezeit. Auch nach z.B. einer abgelaufenen, minimalen Ladezeit will ich ja nicht, dass er bei jedem Sprung unter 1400 Watt (z.B. eine Herdplatte geht immer mal für ein paar Sekunden an, wenn man was aufwärmt), gleich wieder abschaltet (Was er allerdings auch nicht unmittelbar zu tun scheint, er wartet immer bestimmte Zeiten ab, bevor er z.B. wieder ausschaltet etc., allerdings wie genau, habe ich nicht verstanden ). 
Ich möchte das aber selber kontrollieren und möchte so wenig, wie möglich, das Laden an- und ausschalten, daher habe ich in Home Assistant was eigenes angefangen zu Programmieren (Grob gesagt, ich berechne ein Mittel aus den letzten x Minuten und dem GoE wird dann z.B. so lange noch 1400 Watt angezeigt, bis auch das Mittel wieder drunter gefallen ist). 
Im Moment ist allerdings Testen so gut wie unmöglich bei dem trüben Wetter.

**Reply by protree** [reply]
Okay interessant. Ja sehe ich ähnlich. Ich habe eine ähnliche Logik in iobroker programmiert. Ich sende der Wallbox auch Fakewerte um die Ladelogik in meinem Sinne anzupassen. Meine Logik ist grob wie folgt:

Ich habe in meinem Script intern 4 Modi zum PV-Laden definiert, welche abhängig vom SoC meines Hausspeichers geschaltet werden:

stop: Unter 80% SoC des Hausspeichers lädt das Auto gar nicht im PV Modus
Normal: Die echten Leistungswerte der PV werden an die Wallbox gesendet
pvBattLaden: Ab 80% bis 90% SoC des Hausspeichers wird etwas Leistung für das weitere Laden des Hausspeichers genutzt
pvBattEntladen: Zwischen 98% und 90% SoC des Hausspeichers wird etwas Leistung vom Hausspeicher getragen.

Das ganze unterliegt einer Hysterese: Scheint die Sonne, steigt der SoC des Hausspeichers erst auf 90% bis der Modus von Stop auf Normal geschaltet wird und das Auto anfängt zu laden. Fällt der SoC des Hausspeichers auf 80% wird der Modus pvBattLaden so lange aktiviert, bis 90% wieder erreicht sind, dann wird wieder in Normal gewechselt. Steigt der SoC des Hausspeichers auf 98% wird der Modus pvBattEntladen so lange aktiviert bis wieder 90% erreicht sind. In jedem Modus (außer Stop) sende ich der Wallbox solche Leistungswerte, dass sie mit minimaler Ladeleistung weiterlädt und erst aufhört, wenn 80% SoC Hausspeicher unterschritten werden.

Die Phasenumschaltung zwischen 1- und 3-phasigem laden schalte ich manuell durch mein Script. Der Wechsel erfolgt nur an den Grenzen 80% und 98% (SoC Hausspeicher). Beispiele: Das Auto lädt 1-phasig mit 16A also 3,6 kW. Die PV liefert 6kW. Der SoC des Hausspeichers steigt langsam an. Bei 98% SoC Hausspeicher schalte ich auf 3-phasig um und der Modus pvBattEntladen wird aktiv. Das Auto lädt mit etwas über 6 kW und der Hausspeicher entlädt sich langsam auf 90% SoC. Fällt im Tagesverlauf die pvLeistung so lange ab, dass trotz Laden mit 6A 3-phasig (4,2 kW) 80% SoC des Hausspeichers erreicht werden wird dann erst auf 1phasig gewechselt und der Hausspeicher nähert sich langsam wieder den 90% (oder gegen Abend fällt der SoC trotz 1-phasig 6A auf unter 80% und das Laden wird beendet).

Ich nehme somit meinen SoC des Hausspeichers als Regelgröße und habe einen Bereich definiert, in welchem dieser sich bewegen darf (80% - 98%). Das Laden wird in diesem Bereich nicht pausiert und es erfolgt keine Phasenumschaltung, die interne Logik der Wallbox passt aber den Ladestrom zwischen 6A - 16A an. Ich konnte das Script im Oktober gut testen, es hat Wolkendurchgänge von bis zu einer halben Stunde und mehr gut abgepuffert und kaum zwischen 1- und 3-phasigem Laden gewechselt (1-2 mal pro Tag). Der SoC des Hausspeichers blieb ziemlich konstant bei 90% und fiel erst gegen Abend auf 80%. Dadurch konnte ich die Relais-Schaltungen der Wallbox stark minimieren.

Ich kann dir gerne mein Script schicken, ist zwar in Javascript geschrieben aber vlt. kannst du die Logik auch für dich anwenden.

**Reply by GuraxUbrax** [reply]
Ja, den Hausspeicher als Puffer zu verwenden ist natürlich eine tolle Idee, aber ich habe leider keinen. Bei mir soll quasi das Netz puffern bis zu einem gewissen Grad, denn das Auto zu schonen ist mir wichtiger als jeden Spitzenwert einzufangen. 
In Home Assistant geht da auch ein großer Teil ohne Programmierung mit den sog. "Helfer"-Variablen, die man erstellen kann, was ich auch gemacht habe (Zwei Mittelwerte über 7 und 20 Minuten, dann ein Minimum über beide über 15 Minuten und zum Schluss eine Template-Variable für die Berechnung der GoE Werte). Die einzelnen Werte lassen sich dann auch grafisch schön darstellen und man sieht, wie zuverlässig das ganze Arbeitet. 
Aber im Moment fehlt ein wenig die Sonne um weiter zu testen...
